src/utils/allocators.zig

```diff
@@ -0,0 +1,114 @@
+const std = @import("std");
+
+pub const RecycleFBA = struct {
+    // this allocates the underlying memory + dynamic expansions
+    // (only used on init/deinit + arraylist expansion)
+    backing_allocator: std.mem.Allocator,
+    // this does the data allocations (data is returned from alloc)
+    alloc_allocator: std.heap.FixedBufferAllocator,
+    // recycling depot
+    records: std.ArrayList(Record),
+
+    // for thread safety
+    mux: std.Thread.Mutex = .{},
```

Comment:
in the scenario where if its safe to use without locking, they can use it without having an extra lock in the code

Reply:
It just feels like it breaks the allocator abstraction. Normally you can make some assumptions about an allocator, but you can't do that here. There is this extra special knowledge you need to have to ensure you're using it correctly. It feels like it could be easy to misuse. It's not a big deal in this case because you're using it correctly in the one case it was designed for. But I feel like it would be cleaner to integrate this into the allocator somehow. For example you could have some way to optionally specify whether or not you want a thread safe allocator when you instantiate it. The thread safe one would always call the locks internally, and the non-thread safe one wouldn't.

Reply:
yeah thats a great idea - will add a comment to add a config struct which specifies thread-safe version or not